selectors/selectors.md

```diff
@@ -108,3 +108,271 @@ These components imply or expand into the following things:
+
+This use case is inspired by CRDTs, where you have a chain of changes. You observe a new change and want to get all the previous changes up to the one that you have already observed. It is a recursive query with a CID as stopping condition.
+
+The shape of a change could look like this (in JSON):
+
+```json
+{
+  "prev": "prevcid",
+  "timestamp": 1549641260,
+  "value": "abc"
+}
+```
+
+It will be a Recursive Selector following a long a certain field (`prev` in this case):
+
+```json
+{
+  "cidRootedSelector": {
+    "root": "cidabcdef",
+    "selectors": [
+      {"selectCid": "cidabcdef"},
+      {"selectRecursive": {
+        "follow": [
+          {"selectPath": "prev"}
+        ],
+        "cidLimit": "somecid"
+      }}
+    ]
+  }
+}
+```
+
+
+### Getting a full sub-DAG
+
+For getting a full file from UnixFSv1 you need to retrieve a full sub-DAG.
+
+An example selector to get the full sub-DAG rooted at a certain CID:
+
+
+```json
+{
+  "cidRootedSelector": {
+    "root": "cidabcdef",
+    "selectors": [
+      {"selectRecursive": {
+        "follow": [
+          {"selectPath": "Links"},
+          {"selectArrayAll": null},
+          {"selectPath": "multihash"}
+        ]
+      }}
+    ]
+  }
+}
+```
+
+If it's a file in some directory, you can also start at a deeper level:
+
+```json
+{
+  "cidRootedSelector": {
+    "root": "cidabcdef",
+    "selectors": [
+      {"selectCid": "cidabcdef"},
+      {"selectPath": "with"},
+      {"selectPath": "some"},
+      {"selectPath": "subdirectory"},
+      {"selectRecursive": {
+        "follow": [
+          {"selectPath": "Links"},
+          {"selectArrayAll": null},
+          {"selectPath": "multihash"}
+        ]
+      }}
+    ]
+  }
+}
+```
+
+IPLD Schema
+-----------
+
+```ipldsch
+# This is the main entry point for the current selectors
+type CidRootedSelector struct {
+  root Cid
+  # Each element matches a path segement (or several, in some cases)
+  selectors [Selector]
+}
+
+# A catch all for all types of selectors. They are split between recursive and
+# non-recursive ones as a recursive selector can't have another recursive
+# selector embedded
+type Selector union {
+  | SelectPath "selectPath"
+  | SelectArrayAll "selectArrayAll"
+  | SelectArrayPosition "selectArrayPosition"
+  | SelectArraySlice "selectArraySlice"
+  | SelectMapAll "selectMapAll"
+  | SelectRecursive "selectRecursive"
+} representation keyed
+
+# This is a subset of the selectors that can be used within a recursive selector
+type SelectorNonRecursive union {
+  | SelectPath "selectPath"
+  | SelectArrayAll "selectArrayAll"
+  | SelectArrayPosition "selectArrayPosition"
+  | SelectArraySlice "selectArraySlice"
+  | SelectArrayAll "selectMapAll"
+} representation keyed
+
+# Paths are split into their individual segments
+type PathSegement String
+
+# Selects a specific full path
+type SelectPath PathSegment
+
+# Selects all elements of an array, it's similar to a `/*`
+type SelectArrayAll Null
```

Comment:
Couldn't this just be rolled up in to one of the other two array types? Maybe position with a `-1` or slice without bounds specified.

Reply:
Further on the map types: this is likely my ignorance speaking, but is it strictly necessary to treat them different to other properties for the purposes of selecting or is it a matter of needing strict typing for values? You could do:

* array positions as paths/keys, just using integers as your selector value
* "all" for both map and array are essentially the same
* slice is distinct to array types

Reply:
Having `SelectArrayAll` is more explicit (which I like), it also plays nice with type systems.

Have a separate type for "Map" helps for better error checking/messages. You should normally know if it's an array or a map. Though it might make sense to add a "I don't know/care" in the future if it turns out it makes sense.